cou/steps/plan.py

```diff
@@ -172,6 +174,138 @@ def determine_upgrade_target(analysis_result: Analysis) -> OpenStackRelease:
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    """
+    if cli_machines := args.machines:
+        verify_data_plane_cli_machines(cli_machines, analysis_result)
+
+    elif cli_hostnames := args.hostnames:
+        verify_data_plane_cli_hostnames(cli_hostnames, analysis_result)
+
+    elif cli_azs := args.availability_zones:
+        verify_data_plane_cli_azs(cli_azs, analysis_result)
+
+
+def verify_data_plane_cli_machines(cli_machines: list[str], analysis_result: Analysis) -> None:
+    """Verify if the machines passed from the CLI are valid.
+
+    :param cli_machines: Machines passed to the CLI as arguments
+    :type cli_machines: list[str]
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    """
+    data_plane_membership_check(
+        all_options=set(analysis_result.machines.keys()),
+        data_plane_options=set(analysis_result.data_plane_machines.keys()),
+        cli_input=parametrize_cli_inputs(cli_machines),
+        parameter_type="Machine(s)",
+    )
+
+
+def verify_data_plane_cli_hostnames(cli_hostnames: list[str], analysis_result: Analysis) -> None:
+    """Verify if the hostnames passed from the CLI are valid.
+
+    :param cli_hostnames: Hostnames passed to the CLI as arguments
+    :type cli_hostnames: list[str]
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    """
+    all_hostnames = {machine.hostname for machine in analysis_result.machines.values()}
+    data_plane_hostnames = {
+        machine.hostname for machine in analysis_result.data_plane_machines.values()
+    }
+
+    data_plane_membership_check(
+        all_options=all_hostnames,
+        data_plane_options=data_plane_hostnames,
+        cli_input=parametrize_cli_inputs(cli_hostnames),
+        parameter_type="Hostname(s)",
+    )
+
+
+def verify_data_plane_cli_azs(cli_azs: list[str], analysis_result: Analysis) -> None:
+    """Verify if the availability zones passed from the CLI are valid.
+
+    :param cli_azs: AZs passed to the CLI as arguments
+    :type cli_azs: list[str]
+    :param analysis_result:  Analysis result
+    :type analysis_result: Analysis
+    :raises DataPlaneMachineFilterError: When the cloud does not have availability zones.
+    """
+    all_azs_uncast: set[Optional[str]] = {
+        machine.az for machine in analysis_result.machines.values()
+    }
+    data_plane_azs_uncast: set[Optional[str]] = {
+        machine.az for machine in analysis_result.data_plane_machines.values()
+    }
+
+    # NOTE(gabrielcocenza) simple deployments may not have AZs. mypy does not understand
+    # that None is being removed from set, so the type goes from set[Optional[str]] to
+    # set[str]. As we can be sure of that, we can cast with the correct type to mypy.
+    all_azs_uncast.discard(None)
+    data_plane_azs_uncast.discard(None)
+
+    all_azs = cast(set[str], all_azs_uncast)
+    data_plane_azs = cast(set[str], data_plane_azs_uncast)
```

Comment:
Perhaps it's better to also handle this in COUModel like how you are getting the machines?